src/backend/distributed/utils/attribute.c

```diff
@@ -0,0 +1,613 @@
+			colocationGroupId *= 10;
+			colocationGroupId += *tenantEnd - '0';
+			tenantEnd--;
+		}
+
+		int t = colocationGroupId;
+		colocationGroupId = 0;
+		while (t)
+		{
+			colocationGroupId *= 10;
+			colocationGroupId += t % 10;
+			t /= 10;
+		}
+
+		/* hack to get a clean copy of the tenant id string */
+		char tenantEndTmp = *tenantEnd;
+		*tenantEnd = '\0';
+		tenantId = pstrdup(tenantId);
+		*tenantEnd = tenantEndTmp;
+
+		if (MultiTenantMonitoringLogLevel != CITUS_LOG_LEVEL_OFF)
+		{
+			ereport(NOTICE, (errmsg("attributing query to tenant: %s",
+									quote_literal_cstr(tenantId))));
+		}
+
+		strcpy_s(attributeToTenant, sizeof(attributeToTenant), tenantId);
+		attributeToTenantStart = clock();
+	}
+	else
+	{
+		strcpy_s(attributeToTenant, sizeof(attributeToTenant), "");
+	}
+
+	/*DetachSegment(); */
+}
+
+
+/*
+ * AnnotateQuery annotates the query with tenant attributes.
+ */
+char *
+AnnotateQuery(char *queryString, char *partitionColumn, int colocationId)
+{
+	if (partitionColumn == NULL)
+	{
+		return queryString;
+	}
+	StringInfo newQuery = makeStringInfo();
+	appendStringInfo(newQuery, ATTRIBUTE_STRING_FORMAT, partitionColumn, colocationId);
+
+	appendStringInfoString(newQuery, queryString);
+
+	return newQuery->data;
+}
+
+
+/*
+ * CitusAttributeToEnd keeps the statistics for the tenant and calls the previously installed end hook
+ * or the standard executor end function.
+ */
+void
+CitusAttributeToEnd(QueryDesc *queryDesc)
+{
+	/*
+	 * At the end of the Executor is the last moment we have to attribute the previous
+	 * attribution to a tenant, if applicable
+	 */
+	AttributeMetricsIfApplicable();
+
+	/* now call in to the previously installed hook, or the standard implementation */
+	if (prev_ExecutorEnd)
+	{
+		prev_ExecutorEnd(queryDesc);
+	}
+	else
+	{
+		standard_ExecutorEnd(queryDesc);
+	}
+}
+
+
+/*
+ * CompareTenantScore is used to sort the tenant statistics by score
+ * in descending order.
+ */
+static int
+CompareTenantScore(const void *leftElement, const void *rightElement)
+{
+	const TenantStats *leftTenant = (const TenantStats *) leftElement;
+	const TenantStats *rightTenant = (const TenantStats *) rightElement;
+
+	if (leftTenant->score > rightTenant->score)
+	{
+		return -1;
+	}
+	else if (leftTenant->score < rightTenant->score)
+	{
+		return 1;
+	}
+	return 0;
+}
+
+
+/*
+ * AttributeMetricsIfApplicable updates the metrics for current tenant's statistics
+ */
+static void
+AttributeMetricsIfApplicable()
+{
+	if (strcmp(attributeToTenant, "") != 0)
+	{
+		clock_t end = { 0 };
+
+		end = clock();
+		time_t queryTime = time(0);
+		double cpu_time_used = ((double) (end - attributeToTenantStart)) / CLOCKS_PER_SEC;
+
+		if (MultiTenantMonitoringLogLevel != CITUS_LOG_LEVEL_OFF)
+		{
+			ereport(NOTICE, (errmsg("attribute cpu counter (%f) to tenant: %s",
+									cpu_time_used, attributeToTenant)));
+		}
+
+		MultiTenantMonitor *monitor = GetMultiTenantMonitor();
+
+		LWLockAcquire(&monitor->lock, LW_EXCLUSIVE);
```

Comment:
This is a really broad lock that we shouldn't acquire exclusively when not necessary. Now we acquire it for every update of the stats, effectively making the finer grained per-tenant lock completely unused. We should only take this lock when modifying the order of the tenants.

Reply:
SHARE_LOCK
find tenants
if found tenants -> modify tenant
else
    release lock
    acquire EXCLUSIVE_LOCK
    find not find tenanants -> create tenant
    release ECLUSIVE LOCK
    SHARE_LOCK
    find tenant
    if found tenant -> modify tenant
    else -> stop

Reply:
After discussing with @marcocitus: It's probably nice to only sort the array in the exclusive lock some percentage of the time, for the many tenants case (e.g. deviceids). This can be done either probabilistically, or based on time since last taking the lock to create a tenant, or a combination of both. So e.g. `if (lastAdditionByThisProcess > 5 seconds ago || random() > 0.001)`. Not necessarily now, more as something to try out if benchmarks show perf issues.